### ISSUE TITLE
KotlinJsonAdapter can get confused by parameter vs. property names

### ISSUE THREAD

**Issue by cbruegg** [square/moshi#293]
Consider the following test, which fails with `java.lang.AssertionError: Expected <1>, actual <2>.`:

```kotlin
open class NonValParamSuper(val a: Int)
class NonValParam(a: Int, notA: Int = 2): NonValParamSuper(a = notA)

@Test fun kotlinNonValParam() {
  val moshi = Moshi.Builder().add(KotlinJsonAdapterFactory).build()
  val adapter = moshi.adapter(NonValParam::class.java)

  val json = """{"a": 1, "b": 2}"""
  val deserialized = adapter.fromJson(json)
  assertEquals(1, deserialized.a)
}
```
The reason why this happens is because the `KotlinJsonAdapter` [assumes](https://github.com/square/moshi/blob/cd1542363db05ba00a53ec354d1f969fd997a0ab/kotlin/src/main/java/com/squareup/moshi/KotlinJsonAdapter.kt#L176) that the constructor parameter name equals the property name, which isn't necessarily true. This may be a little of an edge case, but I thought it would be worth sharing.

When the `KotlinJsonAdapter` is documented, it should probably contain a warning about this. Another option would be asking JetBrains whether they could extend Kotlin reflection to allow access to an associated property of a constructor parameter in case it has one.

**Comment by swankjesse** [comment]
Yeah, I think this important to track, but unlikely to be a problem in practice.

We should probably confirm that the types match.

**Comment by swankjesse** [comment]
@NightlyNexus any further action on this?
